src/tools/rbd_mirror/Replayer.cc

```diff
@@ -44,50 +93,10 @@ Replayer::~Replayer()
   }
 }
 
-int Replayer::init()
+void Replayer::init(Context *on_finish)
 {
-  dout(20) << "replaying for " << m_peer << dendl;
-
-  int r = m_remote->init2(m_peer.client_name.c_str(),
-			  m_peer.cluster_name.c_str(), 0);
-  if (r < 0) {
-    derr << "error initializing remote cluster handle for " << m_peer
-	 << " : " << cpp_strerror(r) << dendl;
-    return r;
-  }
-
-  r = m_remote->conf_read_file(nullptr);
-  if (r < 0) {
-    derr << "could not read ceph conf for " << m_peer
-	 << " : " << cpp_strerror(r) << dendl;
-    return r;
-  }
-
-  r = m_remote->connect();
-  if (r < 0) {
-    derr << "error connecting to remote cluster " << m_peer
-	 << " : " << cpp_strerror(r) << dendl;
-    return r;
-  }
-
-  dout(20) << "connected to " << m_peer << dendl;
-
-  std::string uuid;
-  r = m_local->cluster_fsid(&uuid);
-  if (r < 0) {
-    derr << "error retrieving local cluster uuid: " << cpp_strerror(r)
-	 << dendl;
-    return r;
-  }
-  m_client_id = uuid;
-
-  // TODO: make interval configurable
-  m_pool_watcher.reset(new PoolWatcher(m_remote, 30, m_lock, m_cond));
-  m_pool_watcher->refresh_images();
-
-  m_replayer_thread.create("replayer");
-
-  return 0;
+  dout(20) << "init replaying for " << m_peer << dendl;
+  m_threads->work_queue->queue(new C_InitReplayer(this, on_finish), 0);
```

Comment:
This doesn't really make this async -- it just makes it synchronous on the work_queue thread.  I don't think it's going to be possible to make the connection to a remote cluster asynchronous.  Given that we should only be talking to 1 or 2 peers, it probably isn't that big of a deal.